pkg/storage/utils/decomposedfs/recycle.go

```diff
@@ -70,92 +71,94 @@ func (fs *Decomposedfs) ListRecycle(ctx context.Context, ref *provider.Reference
 		return nil, errtypes.PermissionDenied(key)
 	}
 
-	spaceID := ref.ResourceId.OpaqueId
 	if key == "" && relativePath == "/" {
 		return fs.listTrashRoot(ctx, spaceID)
 	}
 
-	trashRoot := fs.getRecycleRoot(ctx, spaceID)
-	f, err := os.Open(filepath.Join(trashRoot, key, relativePath))
+	// build a list of trash items relative to the given trash root and path
+	items := make([]*provider.RecycleItem, 0)
+
+	trashRootPath := filepath.Join(fs.getRecycleRoot(ctx, spaceID), lookup.Pathify(key, 4, 2))
+	_, timeSuffix, err := readTrashLink(trashRootPath)
 	if err != nil {
-		if os.IsNotExist(err) {
-			return items, nil
+		sublog.Error().Err(err).Str("trashRoot", trashRootPath).Msg("error reading trash link")
+		return nil, err
+	}
+
+	origin := ""
+	// lookup origin path in extended attributes
+	if attrBytes, err := xattr.Get(trashRootPath, xattrs.TrashOriginAttr); err == nil {
+		origin = string(attrBytes)
+	} else {
+		sublog.Error().Err(err).Str("space", spaceID).Msg("could not read origin path, skipping")
+		return nil, err
+	}
+
+	// all deleted items have the same deletion time
+	var deletionTime *types.Timestamp
+	if parsed, err := time.Parse(time.RFC3339Nano, timeSuffix); err == nil {
+		deletionTime = &types.Timestamp{
+			Seconds: uint64(parsed.Unix()),
+			// TODO nanos
 		}
-		return nil, errors.Wrapf(err, "tree: error listing %s", trashRoot)
+	} else {
+		sublog.Error().Err(err).Msg("could not parse time format, ignoring")
 	}
-	defer f.Close()
 
-	parentNode, err := os.Readlink(filepath.Join(trashRoot, key))
+	trashItemPath := filepath.Join(trashRootPath, relativePath)
+
+	f, err := os.Open(trashItemPath)
 	if err != nil {
-		log.Error().Err(err).Str("trashRoot", trashRoot).Msg("error reading trash link, skipping")
-		return nil, err
+		if os.IsNotExist(err) {
```

Comment:
```suggestion
		if errors.Is(err, fs.ErrNotExist) {
```
`os.IsNotExist` is deprecated. Don't forget to import `fs`.